src/NUnitConsole/nunit-console/Utilities/TeamCityServiceMessagePublisher.cs

```diff
@@ -0,0 +1,389 @@
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    using System;
+
+    internal class TeamCityServiceMessagePublisher
+    {
+        private readonly TextWriter _outWriter;
+        private readonly Dictionary<string, string> _refs = new Dictionary<string, string>();
+        private int _blockCounter;
+        private string _rootFlowId;
+
+        public TeamCityServiceMessagePublisher(TextWriter outWriter)
+        {
+            if (outWriter == null)
+            {
+                throw new ArgumentNullException("outWriter");
+            }
+
+            _outWriter = outWriter;
+        }
+
+        public void RegisterMessage(XmlNode message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            var messageName = message.Name;
+            if (string.IsNullOrEmpty(messageName))
+            {
+                return;
+            }
+
+            messageName = messageName.ToLowerInvariant();
+            if (messageName == "start-run")
+            {
+                _refs.Clear();
+                return;
+            }           
+
+            var fullName = message.GetAttribute("fullname");
+            if (string.IsNullOrEmpty(fullName))
+            {
+                return;
+            }
+
+            var id = message.GetAttribute("id");
+            var parentId = message.GetAttribute("parentId");
+            string flowId;
+            if (parentId != null)
+            {
+                // NUnit 3 case
+                string rootId;
+                flowId = TryFindRootId(parentId, out rootId) ? rootId : id;
+            }
+            else
+            {
+                // NUnit 2 case
+                flowId = _rootFlowId;
+            }
+
+            string testFlowId;
+            if (id != flowId && parentId != null)
+            {
+                testFlowId = id;
+            }
+            else
+            {
+                testFlowId = flowId;
+                if (testFlowId == null)
+                {
+                    testFlowId = id;
+                }
+            }
```

Comment:
Exactly, if we have "parentId", it eliminates the need of a separate "flowId", because we can associate a block opening and closing and we can associate it with related suits. Furthermore it is possible to generate in the place without using "id" as a value. But I did not do it for the best "tracebility".